pkg/client/group_mgmt_client.go

```diff
@@ -69,19 +79,29 @@ func newGroupMgmtClient(ipAddress, username, password, apiVersion string, waitOn
 
 	// Create GroupMgmt Client
 	groupMgmtClient := &GroupMgmtClient{
-		URL:       fmt.Sprintf("https://%s:5392/%s", ipAddress, apiVersion),
+		URL:       fmt.Sprintf(groupURIFmtNonTenant, ipAddress, apiVersion),
 		Client:    restyClient,
 		WaitOnJob: waitOnJobs,
 		Username:  username,
 		Password:  password,
+		isTenant:  false, //the default value. Unless otherwise specified in clientOpts
 	}
+
+	for _, opt := range clientOpts {
+		opt(groupMgmtClient)
+	}
+
+	if groupMgmtClient.isTenant {
+		groupMgmtClient.URL = fmt.Sprintf(groupURIFmtTenant, ipAddress, apiVersion)
+	}
+
 	return groupMgmtClient
 }
 
 // NewClient instantiates a new client to communicate with the Nimble group
-func NewClient(ipAddress, username, password, apiVersion string, waitOnJobs bool) (*GroupMgmtClient, error) {
+func NewClient(ipAddress, username, password, apiVersion string, waitOnJobs bool, clientOpts ...ClientOption) (*GroupMgmtClient, error) {
 	// Get resty client
-	groupMgmtClient := newGroupMgmtClient(ipAddress, username, password, apiVersion, waitOnJobs)
+	groupMgmtClient := newGroupMgmtClient(ipAddress, username, password, apiVersion, waitOnJobs, clientOpts...)
```

Comment:
Thank you for making this change to avoid impacts to existing consumers.  It has me thinking, though....

We're creating a mixture of two factory styles.  The convention for functional options is to use With*() functions rather than Is*() functions.  WithTenant() isn't a great choice because acts more like a modifier for the user name rather than its own thing.  The mixture of styles doesn't let us combine the tenant concept with the user concept.

Also, the waitOnJobs flag shouldn't always need to be specified.  It is a ripe candidate for a functional option.  And apiVersion is always "v1" for the forseeable future.

So I'm wondering if it would be better to essentially deprecate the NewClient() function, leaving it unmodified without tenant support, and introduce a new factory that only uses functional options.

There is an issue of what to call the new factory function, since overloading is not allowed and we couldn't call it NewClient().  But since the convention is to name it New<struct-name>(), we could call it NewGroupMgmtClient().

For the option functions, I suggest:

- WithAddress(string) -- Don't specify "IP" address, since a host address is also accepted
- WithUser(string)
- WithTenantUser(string) -- Use this instead of WithUser() for a tenant user; it sets isTenant=true under the covers
- WithPassword(string)
- WithApiVersion(string)
- WithWaitForAsyncJobs()
- WithoutWaitForAsyncJobs()

I'm interested to hear what others think of this proposal.

Reply:
This is a great idea.  One minor tweak I would offer is changing "WithAddress" to "WithHost".  I tend to see Host used more often than Address when dealing with SDKs.

Reply:
+1 i like the idea too

Reply:
Should `NewNsGroupService()` be depreciated too? Right now the examples file calls on `NewNsGroupService()` and initialize it with ip, username, password, etc. If we were going to get rid of these parameters and replace them with optional parameters in the future, shouldn't  `NewNsGroupService()` also be replaced by another function that only accepts optional parameter (just like `NewGroupMgmtClient()`?

Reply:
You are correct.  NewNsGroupService() is the interface that is used by client consumers.  I think we don't need to worry about compatibility for NewClient(); I had the two confused.  (The NewNsGroupService() changes are well hidden in the PR diff.)

So the question is, if we deprecate NewNsGroupService(), what do we name the replacement function?  I don't have a good answer for that.  Any suggestion?

Note that there is also a NewNsGroupService() function in the fakeservice package which should stay in sync.  I think that was missed in the changes so far.

Reply:
I pushed a new commit addressing this change. For now, I use NewNimbleGroupService() as a placeholder name for the function.